DAOS-16591 mgmt, vos, common: Align scm/meta size

sherintg opened the pull request:
The scm and meta sizes for vos are now aligned-up to 16M for pools using phase 2 allocator.

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

wangshilong commented on src/vos/vos_pool.c:
```diff
@@ -1272,6 +1272,26 @@ vos_pool_create_ex(const char *path, uuid_t uuid, daos_size_t scm_sz, daos_size_
+
+	backend = umempobj_get_backend_type();
+	if ((*scm_sz != *meta_sz) && (backend == DAOS_MD_BMEM))
+		backend = DAOS_MD_BMEM_V2;
```
i think we should return failure in this case?

tanabarr replied:
Why, this seems valid to me. It is the standard case for phase 2 where mem_ratio is not 100% and therefore we need to use V2.

wangshilong replied:
i thought that if users configured backend as DAOS_MD_BMEM using environment DAOS_MD_ON_SSD_MODE. I would think we don't support phase2 in this case? As phase1 and Phase2 will be totally different layout.

sherintg replied:
Addressed in the latest commit. Also, added a new tunable DAOS_MD_DISABLE_BMEM_V2 to disable creation of v2 pools.

wangshilong reviewed (commented):
Anyway, just confirming the question, no objections to the PR.

NiuYawei commented on src/vos/vos_pool.c:
```diff
@@ -1272,6 +1272,26 @@ vos_pool_create_ex(const char *path, uuid_t uuid, daos_size_t scm_sz, daos_size_
+	alignsz = umempobj_pgsz(backend);
+
+	*scm_sz = D_ALIGNUP(*scm_sz, alignsz);
+	if (*meta_sz)
```
Looks like control plane may pass zero meta_sz? Then we could mistakenly assume it's BMEM_V2 backend in above code?

BTW, I think we'd move the "max(tca->tca_scm_size / dss_tgt_nr, 1 << 24)" from tgt_create_preallocate() to here and apply it to both scm size and meta size, right?

sherintg replied:
meta_sz check was added assuming that for pmem mode this may not be set. Verified that that both scm_sz and meta_sz is set by the control plane in all modes pmem, bmem_v1 and bmem_v2. Hence removed the checks and added asserts.

sherintg commented on src/vos/vos_pool.c:
```diff
@@ -1272,6 +1272,26 @@ vos_pool_create_ex(const char *path, uuid_t uuid, daos_size_t scm_sz, daos_size_
+	size_t alignsz;
+
+	backend = umempobj_get_backend_type();
+	if ((*scm_sz != *meta_sz) && (backend == DAOS_MD_BMEM))
```
Addressed

tanabarr replied:
meta and scm sz should always be non-zero as per `src/mgmt/srv_drpc.c L504`

tanabarr commented:
@sherintg please try to avoid force push if you can because it forces reviewers to go through the whole patch again not just the differential. TIA

tanabarr commented on src/mgmt/srv_target.c:
```diff
@@ -1048,15 +1048,14 @@ tgt_create_preallocate(void *arg)
-			rc = tgt_vos_preallocate_sequential(tca->tca_ptrec->dptr_uuid,
-							    max(tca->tca_scm_size / dss_tgt_nr,
-								1 << 24), dss_tgt_nr);
+			rc = tgt_vos_preallocate_sequential(
```
Why do we have to allocate sequentially in MD mode rather than parallel like in PMem?

sherintg replied:
Its the other way round, its sequential in PMEM and parallel in MD. This change was done in phase 1 to overcome the overhead fallocate on tmpfs.

tanabarr commented on src/mgmt/srv_target.c:
```diff
@@ -1123,7 +1122,12 @@ ds_mgmt_hdlr_tgt_create(crt_rpc_t *tc_req)
 	tgt_scm_sz  = tc_in->tc_scm_size / dss_tgt_nr;
 	tgt_meta_sz = tc_in->tc_meta_size / dss_tgt_nr;
-	vos_pool_roundup_size(&tgt_scm_sz, &tgt_meta_sz);
+	rc          = vos_pool_roundup_size(&tgt_scm_sz, &tgt_meta_sz);
```
Is this strange formatting enforced by the clang linter?

sherintg replied:
yes

sherintg commented:
@tanabarr Apologies for doing the force push against second commit.  I did a second commit yesterday addressing the review comments. However, post md_on_ssd meeting I modified this commit (with a force push) to include a tunable to disable phase 2 allocator. The second commit should still contain the difference of what you had reviewed and approved and the changes done as part of rework.

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15146/4/execution/node/1463/log

NiuYawei commented on src/vos/vos_pool.c:
```diff
@@ -753,6 +753,50 @@ init_umem_store(struct umem_store *store, struct bio_meta_context *mc)
+	size_t alignsz;
+	int    rc;
+
+	D_ASSERT((*scm_sz != 0) && (*meta_sz != 0));
```
This assert is triggered in CI tests, looks we pass 0 meta_size sometimes (for pmem or phase1 mode).

sherintg replied:
Addressed with the 3rd commit.

tanabarr replied:
From https://build.hpdd.intel.com/job/daos-stack/job/daos/view/change-requests/job/PR-15146/4/artifact/Functional%20Hardware%20Medium/daos_test/rebuild.py/daos_logs.wolf-254/01-DAOS_Rebuild_0to10_daos_control.log/ failed we can see that the assert happens before pool create, just after engine start (before any `dmg pool` commands are issued).

- Wolf-254 engine 1 starts - rank 2
- Wolf-254 engine 0 starts - rank 5
- SmdQuery issued from dmg to wolf-254 for ranks 2,5
- ranks 2&5 fail due to assert
e.g. `13:06:12 INFO |     Sep 20 13:03:37 wolf-254.wolf.hpdd.intel.com daos_server[167095]: ERROR: daos_engine:0 09/20-13:03:37.29 wolf-254 DAOS[168521/0/4058] vos  EMRG src/vos/vos_pool.c:787 vos_pool_roundup_size() Assertion '(*scm_sz != 0) && (*meta_sz != 0)' failed`

So whilst `meta_sz` will be set nonzero in control-plane pool create calls, other engine internal calls need to be handled for `meta_sz == 0`.
